src/imitation/algorithms/sqil.py

```diff
@@ -0,0 +1,229 @@
+"""
+
+from typing import Any, Dict, List, Optional, Type, Union
+
+import numpy as np
+import torch as th
+from gym import spaces
+from stable_baselines3 import dqn
+from stable_baselines3.common import buffers, policies, type_aliases, vec_env
+from stable_baselines3.dqn.policies import DQNPolicy
+
+from imitation.algorithms import base as algo_base
+from imitation.data import rollout, types
+from imitation.util import logger, util
+
+
+class SQIL(algo_base.DemonstrationAlgorithm[types.Transitions]):
+    """Soft Q Imitation Learning (SQIL).
+
+    Trains a policy via DQN-style Q-learning,
+    replacing half the buffer with expert demonstrations and adjusting the rewards.
+    """
+
+    expert_buffer: buffers.ReplayBuffer
+
+    def __init__(
+        self,
+        *,
+        venv: vec_env.VecEnv,
+        demonstrations: Optional[algo_base.AnyTransitions],
+        policy: Union[str, Type[DQNPolicy]],
+        custom_logger: Optional[logger.HierarchicalLogger] = None,
+        dqn_kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        """Builds SQIL.
+
+        Args:
+            venv: The vectorized environment to train on.
+            demonstrations: Demonstrations to use for training.
+            policy: The policy model to use (SB3).
+            custom_logger: Where to log to; if None (default), creates a new logger.
+            dqn_kwargs: Keyword arguments to pass to the DQN constructor.
+
+        Raises:
+            ValueError: if `dqn_kwargs` includes a key
+                `replay_buffer_class` or `replay_buffer_kwargs`.
+        """
+        self.venv = venv
+
+        if dqn_kwargs is None:
+            dqn_kwargs = {}
+        # SOMEDAY(adam): we could support users specifying their own replay buffer
+        # if we made SQILReplayBuffer a more flexible wrapper. Does not seem worth
+        # the added complexity until we have a concrete use case, however.
+        if "replay_buffer_class" in dqn_kwargs:
+            raise ValueError(
+                "SQIL uses a custom replay buffer: "
+                "'replay_buffer_class' not allowed.",
+            )
+        if "replay_buffer_kwargs" in dqn_kwargs:
+            raise ValueError(
+                "SQIL uses a custom replay buffer: "
+                "'replay_buffer_kwargs' not allowed.",
+            )
+
+        self.dqn = dqn.DQN(
+            policy=policy,
+            env=venv,
+            replay_buffer_class=SQILReplayBuffer,
+            replay_buffer_kwargs={"demonstrations": demonstrations},
+            **dqn_kwargs,
+        )
+
+        super().__init__(demonstrations=demonstrations, custom_logger=custom_logger)
+
+    def set_demonstrations(self, demonstrations: algo_base.AnyTransitions) -> None:
+        assert isinstance(self.dqn.replay_buffer, SQILReplayBuffer)
+        self.dqn.replay_buffer.set_demonstrations(demonstrations)
+
+    def train(self, *, total_timesteps: int):
+        self.dqn.learn(total_timesteps=total_timesteps)
```

Comment:
`DQN.learn` has a bunch of optional parameters which would be good to expose imo:
```python
        callback: MaybeCallback = None, # this one seems good to have
        log_interval: int = 4,  # ...same here
        tb_log_name: str = "DQN",  # this one too; maybe override the default to sth else than "DQN"?  e.g. "SQIL-DQN" ?
        reset_num_timesteps: bool = True,  # not sure about this one
        progress_bar: bool = False,  # I definitely want to have this one :D
```
maybe just pass through the kwargs?  possibly with a different default value for "tb_log_name"

Reply:
Good point, passing through `**kwargs` seems like a good solution. Changing `tb_log_name` to default to `SQIL` seems reasonable.